vql/server/splunk.go

```diff
@@ -0,0 +1,241 @@
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+/* Plugin Splunk.
+
+
+ */
+package server
+
+import (
+	"context"
+	"crypto/tls"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/Velocidex/ordereddict"
+	"github.com/ZachtimusPrime/Go-Splunk-HTTP/splunk"
+	"www.velocidex.com/golang/velociraptor/acls"
+	vql_subsystem "www.velocidex.com/golang/velociraptor/vql"
+	vfilter "www.velocidex.com/golang/vfilter"
+)
+
+type _SplunkPluginArgs struct {
+	Query      vfilter.StoredQuery `vfilter:"required,field=query,doc=Source for rows to upload."`
+	Threads    int64               `vfilter:"optional,field=threads,doc=How many threads to use."`
+	URL        string              `vfilter:"optional,field=url,doc=The Splunk Event Collector URL."`
+	Token      string              `vfilter:"optional,field=token,doc=Splunk HEC Token."`
+	Index      string              `vfilter:"required,field=index,doc=The name of the index to upload to."`
+	Source     string              `vfilter:"optional,field=source,doc=The source field for splunk. If not specified this will be 'velociraptor'."`
+	Sourcetype string              `vfilter:"optional,field=sourcetype,doc=The sourcetype field for splunk. If not specified this will 'vql'"`
+	ChunkSize  int64               `vfilter:"optional,field=chunk_size,doc=The number of rows to send at the time."`
+	SkipVerify bool                `vfilter:"optional,field=skip_verify,doc=Skip SSL verification(default: False)."`
+
+	WaitTime int64 `vfilter:"optional,field=wait_time,doc=Batch splunk upload this long (2 sec)."`
+}
+
+type _SplunkPlugin struct{}
+
+func (self _SplunkPlugin) Call(ctx context.Context,
+	scope *vfilter.Scope,
+	args *ordereddict.Dict) <-chan vfilter.Row {
+	output_chan := make(chan vfilter.Row)
+
+	go func() {
+		defer close(output_chan)
+
+		err := vql_subsystem.CheckAccess(scope, acls.COLLECT_SERVER)
+		if err != nil {
+			return
+		}
+
+		arg := _SplunkPluginArgs{}
+		err = vfilter.ExtractArgs(scope, args, &arg)
+		if err != nil {
+			return
+		}
+
+		if arg.Threads == 0 {
+			arg.Threads = 1
+		}
+
+		if arg.ChunkSize == 0 {
+			arg.ChunkSize = 1000
+		}
+
+		if arg.WaitTime == 0 {
+			arg.WaitTime = 2
+		}
+
+		if len(arg.Sourcetype) == 0 {
+			arg.Sourcetype = "vql"
+		}
+
+		if len(arg.Source) == 0 {
+			arg.Source = "velociraptor"
+		}
+
+		wg := sync.WaitGroup{}
+		row_chan := arg.Query.Eval(ctx, scope)
+		for i := 0; i < int(arg.Threads); i++ {
+			wg.Add(1)
+
+			// Start an uploader on a thread.
+			go _upload_rows(ctx, scope, output_chan,
+				row_chan, &wg, &arg)
+		}
+
+		wg.Wait()
+	}()
+	return output_chan
+}
+
+// Copy rows from row_chan to a local buffer and push it up to splunk.
+func _upload_rows(
+	ctx context.Context,
+	scope *vfilter.Scope, output_chan chan vfilter.Row,
+	row_chan <-chan vfilter.Row,
+	wg *sync.WaitGroup,
+	arg *_SplunkPluginArgs) {
+	defer wg.Done()
+
+	var buf []*ordereddict.Dict
```

Comment:
Maybe create the slice with the chunk capacity to start with? This avoid re-allocs

```
make([]*ordereddict.Dict, 0, arg.ChunkSize)
```